ilastik/shell/gui/reportIssueDialog.py

```diff
@@ -0,0 +1,94 @@
+import os
+
+import ilastik
+import platform
+import sys
+import webbrowser
+from PyQt5.QtWidgets import QApplication, QDialog, QVBoxLayout, QLabel, QTextEdit, QPushButton, QHBoxLayout, QSizePolicy
+from urllib.parse import quote
+
+
+def _get_log_since_last_startup() -> str:
+    """
+    Reads the log up to maximum 1 MB from the end and returns the text since the last startup, or the entire 1 MB.
+    We don't want people emailing us more than 1MB of log...
+    """
+    log_path = ilastik.ilastik_logging.default_config.get_logfile_path()
+    marker = b"Starting ilastik"
```

Comment:
This is dangerous; It attributes logical significance to a throwaway `log` statement elsewhere in the code, which could also be disabled via logging configuration. You could make this official, by putting the `"Starting ilastik"` string in a constant somewhere, commenting heavily that such string is not to be touched, and then reference such constant both in the print statement as well as in this function.

Still there's the problem that this string isn't supper unique; Something logging a phrase like `[info] Starting ilastik with such and such parameters will make it do so and so...` would throw this logic off.

And finally, `marker` is looking for the ASCII representation of the string `"Starting ilastik"`. Are we sure windows isn't writing that string in some `UTF16-like` encoding that will escape detection when you do `marker in log_text` ? If that turns out to be the case, it'd be safer (though slower) to first decode the bytes, then search within the string instead of the raw bytes.

Maybe we could try something safer, like recording the initial size of the log file on startup and then just grab everything that comes after that? Though I haven't thought deeply about how this could fail

Reply:
+1 for string constant. I am on Windows and the encoding was no issue, but +1 for decoding before comparison. I don't think decoding 500kib of text is worth optimising away here.

I'd much rather risk us screwing up the signal at some point in the future than tie this dialog (which I'm trying not to over-engineer) to a global state defined at startup...

Reply:
> than tie this dialog [...] to a global state defined at startup

A file that is being written elsewhere **is** global state. Even more so when there's the expectation of a marker showing up on it.

Another issue is that we don't know how much buffering the logging system might be using. For all we know at this point in the code, it might just be using `io.DEFAULT_BUFFER_SIZE`, so we might miss the last `~8192` bytes of logging that just haven't been flushed down to the file yet.

I was thinking  about the crash detection idea from last meeting. If the temporary file that is created for crash detection is also that run's log file, we could just take the entire contents of that file; no reading from the end, no looking for markers :thinking:

Reply:
I don't see a problem here: If the marker doesn't show up, worst case we read 64kib of text. If the logger hasn't flushed in a while, worst case we miss some log. Should I add an explicitly flush of the RotatingFileHandler before we read the log file?

The only real way I see to make anything here more elegant would be one-log-per-session, but I would leave that for another time.

Reply:
> Should I add an explicitly flush of the RotatingFileHandler [...]?

That would help. Still, the log could have recently rotated, which would cause other problems.

The way I see it, our bug reporter can't be itself buggy. If we end up deciding that we can't trust the submitted logs and would always ask the user to send the entire thing to us, then we waste the effort of this feature. Some still standing problems with this approach are:

- The log rotated and chopped the log;
- The extra stuff we get before the marker contains another marker which confuses our human parsing;
- someone somewhere logs something that is detected as a marker;
- reconfiguring the logger invalidates our assumptions;
- The clipped log at the start could hinder our anonymizing efforts by making strings less path-like

Some of these could be ameliorated by reading the entire things as a string instead of bytes (I've pointed byte encoding issue elsewhere) , and maybe doing something more line-oriented. The log rotation and reconfiguration could still get us, though.

Reply:
I think these should all be addressed now with the session log. The downside is that the only way to implement this without breaking whatever people might already be doing with our logs, is to also keep the old config and double log everything. I'd be happy to go more aggressive on the log file limit for the session log (i.e. rotate earlier, keep fewer chunks), since the config for the main file (20MB, 5 chunks) guarantees it can't grow above 100MB in total, while for session logs there is no limit to their number. This would make https://github.com/ilastik/ilastik/issues/2717 a bit more urgent.

> The log rotated and chopped the log;

The most recent chunk retains the "main" log file name, so that gets read in this case. Not sure how this could be an issue.

> The extra stuff we get before the marker contains another marker which confuses our human parsing;

Gone

> someone somewhere logs something that is detected as a marker;

Gone

> reconfiguring the logger invalidates our assumptions;

The user gets to put their own log path through the `logfile` cmdline parameter. If that's the case, no session log. I think that's the most sensible since the user is more likely to be running things on the files in that folder if they are using a custom one. The dialog falls back to not including any log in that case, and I added an "open log folder" button instead. If they are somehow messing with the logging config beyond that, they're advanced enough not to need our help (or handholding on how to contact us)...

> The clipped log at the start could hinder our anonymizing efforts by making strings less path-like

Gone. I figured this would also be a potential issue if logs can be rotated mid-message, but as expected `logging` is smart enough. If you set the rotation size to 1, it creates one file per message.